pkg/controller/hibernation/hibernation_controller.go

```diff
@@ -148,6 +149,63 @@ func (r *hibernationReconciler) Reconcile(request reconcile.Request) (reconcile.
 		return reconcile.Result{}, nil
 	}
 
+	// Check for the hibernate-after annotation, and if the cluster has been in running state for longer than this duration, put it to sleep.
+	hibernateAfter, ok := cd.Annotations[constants.HibernateAfterAnnotation]
+	if ok && cd.Spec.PowerState != hivev1.HibernatingClusterPowerState {
+		cdLog.Debugf("found hibernate-after annotation: %s", hibernateAfter)
+		hibernateAfterDur, err := time.ParseDuration(hibernateAfter)
+		if err != nil {
+			cdLog.WithError(err).WithField("hibernateAfter", hibernateAfter).Infof("error parsing %s as a duration, ignoring", constants.HibernateAfterAnnotation)
+		}
+
+		runningSince := cd.Status.InstalledTimestamp.Time
+
+		cond := controllerutils.FindClusterDeploymentCondition(cd.Status.Conditions, hivev1.ClusterHibernatingCondition)
+		if cond == nil {
+			cdLog.WithField("runningSince", runningSince).Debug("cluster has no hibernating condition (never hibernated), using installed time")
+		} else if cond.Status == corev1.ConditionFalse {
+			runningSince = cond.LastTransitionTime.Time
+			cdLog.WithField("runningSince", runningSince).WithField("reason", cond.Reason).Debug("hibernating condition false")
+		} else {
+			// Cluster is Hibernating, we have no runningSince time:
+			runningSince = time.Time{}
+		}
+
+		hibLog := cdLog.WithFields(log.Fields{
+			"runningSince":   runningSince,
+			"hibernateAfter": hibernateAfter,
+		})
+
+		if !runningSince.IsZero() {
```

Comment:
Would like to see this done as a separate variable rather than relying on the implicit understanding that a zero value for `runningSince` means that the cluster is hibernating. In other words, create an `isHibernating` variable that is set to true if the cluster is hibernating.